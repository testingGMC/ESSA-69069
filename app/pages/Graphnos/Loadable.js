/**
 *
 * Asynchronously loads the component for Graphnos
 *
 */

import loadable from 'utils/loadable'

export default loadable(() => import('./index'))
