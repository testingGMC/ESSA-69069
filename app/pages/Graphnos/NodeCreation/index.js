import { Form } from 'antd'
import NodeCreation from './NodeCreation'

export default Form.create({ name: 'create_node' })(NodeCreation)
