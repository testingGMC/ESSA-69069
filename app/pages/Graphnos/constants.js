export const colors = {
  0: '#e74c3c', // skill
  1: '#3a9d23', // superSkill
  2: '#2980b9', // track
  3: '#E1A101', // checkpoint
  4: '#800080', // reward
  5: '#FF6633', // workshop
  6: 'blue', // traversed node
}

export const types = {
  '#e74c3c': 0,
  '#3a9d23': 1,
  '#2980b9': 2,
  '#E1A101': 3,
  '#800080': 4,
  '#FF6633': 5,
  blue: 6,
  // '#800080': 7,
  // '#FF6633': 8,
}

export const original = 'original'
export const preview = 'preview'
export const pBro = 'p-bro'
export const reversePBro = 'reverse'
export const magnify = 'magnify'
export const creation = 'creation'
export const update = 'update'

export const nodeTypes = {
  skill: 0,
  superSkill: 1,
  track: 2,
  checkpoint: 3,
  reward: 4,
  workshop: 5,
}
