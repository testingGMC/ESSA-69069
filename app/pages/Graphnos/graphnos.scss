.graphnos{
    margin: auto;
    padding: 35px;
    .ant-page-header {
        display: inline-flex;
        align-items: center;
        padding: 16px 4px;

        .ant-page-header-title-view-title {
            font-size: 28px;
        }
    }
}