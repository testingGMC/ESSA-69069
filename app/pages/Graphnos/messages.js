/*
 * Graphnos Messages
 *
 * This contains all the text for the Graphnos container.
 */

import { defineMessages } from 'react-intl'

export const scope = 'app.containers.Graphnos'

export default defineMessages({
  pageTitle: {
    id: `${scope}.pageTitle`,
    defaultMessage: 'Graphnos',
  },
  helmetDescription: {
    id: `${scope}.helmetDescription`,
    defaultMessage: 'Description of Graphnos',
  },
  helmetTitle: {
    id: `${scope}.helmetTitle`,
    defaultMessage: 'Graphnos',
  },
})
