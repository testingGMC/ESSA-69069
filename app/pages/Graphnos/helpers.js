import { COLORS } from 'shared/constants'

export const findNode = (nodes, nodeId) =>
  nodes.find(node => node.id === nodeId)

export const findLink = (links, source, target) =>
  links.find(
    link =>
      (link.source === source && link.target === target) ||
      (link.source === target && link.target === target),
  )

export const retrieveNodes = (nodes, source, target) => {
  const result = nodes.reduce(
    (acc, node) => {
      if (node.id === source) {
        acc[0] = node
      } else if (node.id === target) {
        acc[1] = node
      }
      return acc
    },
    ['source', 'target'],
  )
  return result
}

const nodesStructTransformation = nodes => {
  const newNodes = {}
  nodes.forEach(node => {
    newNodes[node.id] = { ...node, successors: [], dependencies: [] }
  })
  return newNodes
}

const dependenciesStructTransformation = dependencies => {
  const newDependencies = {}
  dependencies.forEach(dependency => {
    newDependencies[dependency.id] = { ...dependency }
  })
  return newDependencies
}

const fillSuccessors = graph => {
  const newGraph = { ...graph }
  const dependenciesId = Object.keys(newGraph.dependencies)
  dependenciesId.forEach(id => {
    const dependency = newGraph.dependencies[id]
    newGraph.nodes[dependency.sourceId].successors.push(dependency.targetId)
  })
  return newGraph.nodes
}

const fillNodesOutgoingDependencies = graph => {
  const newGraph = { ...graph }
  const dependenciesId = Object.keys(newGraph.dependencies)
  dependenciesId.forEach(id => {
    const dependency = newGraph.dependencies[id]
    newGraph.nodes[dependency.sourceId].dependencies.push(id)
  })
  return newGraph.nodes
}

const fillNodesDependencies = graph => {
  const newGraph = { ...graph }
  const dependenciesId = Object.keys(newGraph.dependencies)
  dependenciesId.forEach(id => {
    const dependency = newGraph.dependencies[id]
    newGraph.nodes[dependency.sourceId].dependencies.push(id)
    newGraph.nodes[dependency.targetId].dependencies.push(id)
  })
  return newGraph.nodes
}

export const newGraphStructDuplicateSkill = graph => {
  const newGraph = {
    nodes: nodesStructTransformation(graph.nodes),
    dependencies: dependenciesStructTransformation(graph.dependencies),
  }
  newGraph.nodes = fillSuccessors(newGraph)
  newGraph.nodes = fillNodesDependencies(newGraph)
  return newGraph
}

export const newGraphStruct = graph => {
  const newGraph = {
    nodes: nodesStructTransformation(graph.nodes),
    dependencies: dependenciesStructTransformation(graph.dependencies),
  }
  newGraph.nodes = fillSuccessors(newGraph)
  newGraph.nodes = fillNodesOutgoingDependencies(newGraph)
  return newGraph
}

// here we assume we have a graph with nodes with
// with the right structure and filled with their
// successors and dependencies (where the node is
// the source)

const extractSubGraphNodes = (initialNode, graph, newNodes) => {
  if (initialNode.successors.length) {
    initialNode.successors.forEach(successor => {
      extractSubGraphNodes(graph.nodes[successor], graph, newNodes)
    })
  }
  // newNodes.push(initialNode)
  newNodes[initialNode.id] = initialNode
}

const extractSubGraphDependencies = (graph, newNodes) => {
  const dependenciesId = Object.keys(graph.dependencies)
  const dependencies = dependenciesId.filter(id => {
    const dependency = graph.dependencies[id]
    return newNodes[dependency.sourceId] && newNodes[dependency.targetId]
  })

  const result = {}

  dependencies.forEach(el => {
    result[el] = graph.dependencies[el]
  })
  return result
}

export const extractSubGraph = (initialNode, graph) => {
  const newNodes = {}
  extractSubGraphNodes(initialNode, graph, newNodes)
  const newDependencies = extractSubGraphDependencies(graph, newNodes)
  return {
    nodes: newNodes,
    dependencies: newDependencies,
  }
}

const sortDependenciesByWeight = (nodeId, graph) => {
  const { dependencies } = graph.nodes[nodeId]
  return dependencies.sort(
    (a, b) => graph.dependencies[a].weight - graph.dependencies[b].weight,
  )
}

const traversal = (initialNode, subGraph, newNodes) => {
  if (initialNode.successors.length) {
    const sortedDependenciesIds = sortDependenciesByWeight(
      initialNode.id,
      subGraph,
    )
    sortedDependenciesIds.forEach(dependency => {
      const node = subGraph.nodes[subGraph.dependencies[dependency].targetId]
      traversal(node, subGraph, newNodes)
    })
  }
  newNodes.push(initialNode)
}

export const traversalSimulation = (initialNode, graph) => {
  const structuredGraph = newGraphStruct(graph)
  const subGraph = extractSubGraph(
    structuredGraph.nodes[initialNode.id],
    structuredGraph,
  )
  const newNodes = []
  traversal(structuredGraph.nodes[initialNode.id], subGraph, newNodes)
  return {
    nodes: newNodes,
    dependencies: Object.keys(subGraph.dependencies).map(
      key => subGraph.dependencies[key],
    ),
  }
}

const hasCircularDependency = (initialNode, graph, sourceNode) => {
  if (initialNode.successors.length) {
    const sortedDependenciesIds = sortDependenciesByWeight(
      initialNode.id,
      graph,
    )
    // eslint-disable-next-line no-restricted-syntax
    for (const dependency of sortedDependenciesIds) {
      const node = graph.nodes[graph.dependencies[dependency].targetId]
      if (node.id === sourceNode.id) return true
      const isCircular = hasCircularDependency(node, graph, sourceNode)
      if (isCircular) return true
    }
  }
  return false
}

export const checkCircularDependency = (sourceNode, targetNode, graph) => {
  const newGraph = {
    ...graph,
    dependencies: [
      ...graph.dependencies,
      {
        sourceId: sourceNode.id,
        targetId: targetNode.id,
        id: sourceNode.id + targetNode.id,
        weight: 1,
      },
    ],
  }
  const structuredGraph = newGraphStruct(newGraph)
  return hasCircularDependency(
    structuredGraph.nodes[targetNode.id],
    structuredGraph,
    sourceNode,
  )
}

export const graphTransformation = graph => {
  const nodes = graph.nodes.map(node => ({
    id: node.id,
    color: COLORS[node.type],
    name: node.name,
    type: node.type,
    description: node.description,
  }))
  const dependencies = graph.dependencies.map(dependency => ({
    source: dependency.sourceId,
    target: dependency.targetId,
    weight: dependency.weight,
    id: dependency.id,
  }))
  return {
    nodes,
    links: dependencies,
  }
}
