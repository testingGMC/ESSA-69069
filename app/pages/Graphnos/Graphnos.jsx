/**
 *
 * Graphnos
 *
 */

import React, { useEffect } from 'react'

import PropTypes from 'prop-types'

import { Helmet } from 'react-helmet'
import { injectIntl, intlShape } from 'react-intl'
import { Switch, Route, Redirect } from 'react-router-dom'
import { isContentArchitect } from 'shared/utils/access-token'
import routes from 'shared/routes'
import FourOfFour from 'pages/FourOfFour/'
import messages from './messages'
// sub pages
import SkillNetwork from './SkillNetwork'
import Traversal from './Traversal'
import Transform from './Transform'
import { traversalSimulation } from './helpers'

import './graphnos.scss'

const redirectSkillNetwork = () => <Redirect to="/graphnos/skill-network" />

const Graphnos = ({
  fetchGraph,
  graphnos,
  deleteNode,
  createNode,
  updateNode,
  createOrUpdateDependency,
  deleteDependency,
  loadSimulatedGraph,
  updateSimulationNode,
  duplicateNode,
  resetSimulation,
  finishSimulation,
  applyPBro,
  applyReversePBro,
  magnify,
  redirectToEditor,
  updateTrackFilters,
  updateNodeFilters,
  loadSimulatedNodes,
  roles,
  intl,
  selectNode,
  setPreviousNode,
  setLink,
  setShowModal,
  clearReducer,
  push,
  connectionStatus
}) => {
  useEffect(
    () => () => {
      clearReducer()
    },
    [],
  )
  useEffect(() => {
    async function fetchGraphWrapper() {
      await fetchGraph()
    }
    fetchGraphWrapper()
  }, [])
  const findNode = (nodeId, graph) =>
    graph.nodes.find(node => node.id === nodeId)
  const subGraphs = []
  const graphToDisplay = {
    nodes: [],
    dependencies: [],
  }
  if (graphnos.data.graph.nodes.length) {
    graphnos.data.trackFilters.forEach(track => {
      subGraphs.push(traversalSimulation(track, graphnos.data.graph))
    })
    graphnos.data.nodeFilters.forEach(node => {
      subGraphs.push({
        nodes: [findNode(node.id, graphnos.data.graph)],
        dependencies: [],
      })
    })
  }

  subGraphs.forEach(subGraph => {
    graphToDisplay.nodes = [...graphToDisplay.nodes, ...subGraph.nodes]
    graphToDisplay.dependencies = [
      ...graphToDisplay.dependencies,
      ...subGraph.dependencies,
    ]
  })
  const helmetMessages = {
    helmetTitle: intl.formatMessage({
      ...messages.helmetTitle,
    }),
    helmetDescription: intl.formatMessage({
      ...messages.helmetDescription,
    }),
  }
  return (
    <div className="graphnos">
      <Helmet>
        <title>{helmetMessages.helmetTitle}</title>
        <meta name="description" content={helmetMessages.helmetDescription} />
      </Helmet>
      <Switch>
        <Route
          exact
          path={routes.GRAPHNOS.linkPath(routes.GRAPHNOS.children.skillNetwork)}
          render={props => (
            <SkillNetwork
              loading={graphnos.loading}
              graphToDisplay={
                graphToDisplay.nodes.length
                  ? graphToDisplay
                  : graphnos.data.graph
              }
              graphData={graphnos.data.graph}
              deleteNode={deleteNode}
              createNode={createNode}
              updateNode={updateNode}
              createOrUpdateDependency={createOrUpdateDependency}
              deleteDependency={deleteDependency}
              duplicateNode={duplicateNode}
              loadSimulatedGraph={loadSimulatedGraph}
              redirectToEditor={redirectToEditor}
              simulatedGraph={graphnos.data.simulatedGraph}
              updateTrackFilters={updateTrackFilters}
              updateNodeFilters={updateNodeFilters}
              nodeFilters={graphnos.data.nodeFilters}
              trackFilters={graphnos.data.trackFilters}
              loadSimulatedNodes={loadSimulatedNodes}
              roles={roles}
              selectedNode={graphnos.data.selectedNode}
              previousNode={graphnos.data.previousNode}
              link={graphnos.data.link}
              selectNode={selectNode}
              setPreviousNode={setPreviousNode}
              setLink={setLink}
              showModal={graphnos.data.showModal}
              setShowModal={setShowModal}
              connectionStatus={connectionStatus}
              {...props}
            />
          )}
        />
        <Route
          exact
          path={routes.GRAPHNOS.linkPath(routes.GRAPHNOS.children.traversal())}
          render={props => (
            <Traversal
              simulatedGraph={graphnos.data.simulatedGraph}
              graphData={graphnos.data.graph}
              loadSimulatedGraph={loadSimulatedGraph}
              updateSimulationNode={updateSimulationNode}
              isSimulating={graphnos.data.isSimulating}
              resetSimulation={resetSimulation}
              finishSimulation={finishSimulation}
              graphToDisplay={
                graphToDisplay.nodes.length
                  ? graphToDisplay
                  : graphnos.data.graph
              }
              push={push}
              {...props}
            />
          )}
        />
        <Route
          exact
          path={routes.GRAPHNOS.linkPath(routes.GRAPHNOS.children.transform)}
          render={props =>
            isContentArchitect(roles) ? (
              <Transform
                applyPBro={applyPBro}
                applyReversePBro={applyReversePBro}
                magnify={magnify}
                graphData={
                  graphToDisplay.nodes.length
                    ? graphToDisplay
                    : graphnos.data.graph
                }
                graphToDisplay={
                  graphToDisplay.nodes.length
                    ? graphToDisplay
                    : graphnos.data.graph
                }
                {...props}
              />
            ) : (
              <Redirect
                to={routes.GRAPHNOS.linkPath(
                  routes.GRAPHNOS.children.skillNetwork,
                )}
              />
            )
          }
        />
        <Route component={FourOfFour} />
      </Switch>
    </div>
  )
}

Graphnos.propTypes = {
  fetchGraph: PropTypes.func.isRequired,
  graphnos: PropTypes.objectOf(PropTypes.any).isRequired,
  deleteNode: PropTypes.func.isRequired,
  createNode: PropTypes.func.isRequired,
  updateNode: PropTypes.func.isRequired,
  createOrUpdateDependency: PropTypes.func.isRequired,
  deleteDependency: PropTypes.func.isRequired,
  loadSimulatedGraph: PropTypes.func.isRequired,
  updateSimulationNode: PropTypes.func.isRequired,
  duplicateNode: PropTypes.func.isRequired,
  resetSimulation: PropTypes.func.isRequired,
  finishSimulation: PropTypes.func.isRequired,
  applyPBro: PropTypes.func.isRequired,
  applyReversePBro: PropTypes.func.isRequired,
  magnify: PropTypes.func.isRequired,
  redirectToEditor: PropTypes.func.isRequired,
  updateTrackFilters: PropTypes.func.isRequired,
  updateNodeFilters: PropTypes.func.isRequired,
  loadSimulatedNodes: PropTypes.func.isRequired,
  roles: PropTypes.array.isRequired,
  intl: intlShape.isRequired,
  selectNode: PropTypes.func.isRequired,
  setPreviousNode: PropTypes.func.isRequired,
  setLink: PropTypes.func.isRequired,
  setShowModal: PropTypes.func.isRequired,
  clearReducer: PropTypes.func.isRequired,
  push: PropTypes.func.isRequired,
  connection: PropTypes.object.isRequired,
  isConnectedToLiveshare: PropTypes.bool.isRequired,
  connectionStatus: PropTypes.number.isRequired,
}

export default injectIntl(Graphnos)
