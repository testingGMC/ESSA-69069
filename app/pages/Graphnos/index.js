/**
 *
 * Graphnos
 *
 */

import React, { memo } from 'react'
import { connect } from 'react-redux'

import { createStructuredSelector } from 'reselect'
import { compose } from 'redux'
import injectSaga from 'utils/injectSaga'
import { DAEMON } from 'utils/constants'
import { useInjectReducer } from 'utils/injectReducer'
import makeSelectGraphnos from 'shared/redux/graphnos/selectors'
import reducer from 'shared/redux/graphnos/reducer'
import liveShareReducer from 'shared/redux/liveshare/reducer'
import * as actions from 'shared/redux/graphnos/actions'
import liveshareSaga from 'shared/redux/liveshare/saga'
import saga from 'shared/redux/graphnos/saga'
import {
  getConnection,
  getConnected,
  getConnectionStatus,
} from 'shared/redux/liveshare/selectors'
import editorReducer from 'shared/redux/editor/reducer'
import editorSaga from 'shared/redux/editor/saga'
import { loadSimulatedNodes } from 'shared/redux/editor/actions'
import { push } from 'connected-react-router'

import Graphnos from './Graphnos'

const GraphnosIndex = props => {
  useInjectReducer({ key: 'graphnos', reducer })
  useInjectReducer({ key: 'editor', reducer: editorReducer })
  useInjectReducer({ key: 'liveshare', reducer: liveShareReducer })
  return <Graphnos {...props} />
}

const mapStateToProps = createStructuredSelector({
  graphnos: makeSelectGraphnos(),
  connection: getConnection(),
  isConnectedToLiveshare: getConnected(),
  connectionStatus: getConnectionStatus(),
})

const mapDispatchToProps = {
  ...actions,
  loadSimulatedNodes,
  push,
}

const withSaga = injectSaga({ key: 'graphnos', saga, mode: DAEMON })
const withSagaEditor = injectSaga({
  key: 'editor',
  saga: editorSaga,
  mode: DAEMON,
})
const withSagaLiveShare = injectSaga({
  key: 'liveshare',
  saga: liveshareSaga,
  mode: DAEMON,
})
const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
)

export default compose(
  withConnect,
  memo,
  withSaga,
  withSagaEditor,
  withSagaLiveShare,
)(GraphnosIndex)
