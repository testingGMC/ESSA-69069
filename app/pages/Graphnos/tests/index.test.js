import React from 'react'
import { render } from 'react-testing-library'
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'
import { browserHistory , BrowserRouter } from 'react-router-dom'

import GraphnosIndex from '../index'
import Graphnos from '../Graphnos'
import { DEFAULT_LOCALE } from '../../../i18n'
import configureStore from '../../../configureStore'

const initialProps = {
  fetchGraph: jest.fn(),
  graphnos: {
    data: {
      graph: {
        nodes: {},
        trackFilters: {}
      }
    }
  },
  deleteNode: jest.fn(),
  createNode: jest.fn(),
  updateNode: jest.fn(),
  clearReducer: jest.fn(),
  createOrUpdateDependency: jest.fn(),
  deleteDependency: jest.fn(),
  loadSimulatedGraph: jest.fn(),
  updateSimulationNode: jest.fn(),
  duplicateNode: jest.fn(),
  resetSimulation: jest.fn(),
  finishSimulation: jest.fn(),
  applyPBro: jest.fn(),
  applyReversePBro: jest.fn(),
  magnify: jest.fn(),
  redirectToEditor: jest.fn(),
  updateTrackFilters: jest.fn(),
  updateNodeFilters: jest.fn(),
  loadSimulatedNodes: jest.fn(),
  roles: '',
  intl: 0,
  selectNode: jest.fn(),
  setPreviousNode: jest.fn(),
  setLink: jest.fn(),
}

describe('<GraphnosIndex />', () => {
  let store
  beforeAll(() => {
    store = configureStore({}, browserHistory)
  })

  it('Expect to not log in console', () => {
    const spy = jest.spyOn(global.console, 'log')
    render(
      <Provider store={store}>
        <BrowserRouter>
          <IntlProvider locale={DEFAULT_LOCALE}>
            <GraphnosIndex {...initialProps} />
          </IntlProvider>
        </BrowserRouter>
      </Provider>
    )
    expect(spy).not.toHaveBeenCalled()
  })

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error')
    render(
      <Provider store={store}>
        <BrowserRouter>
          <IntlProvider locale={DEFAULT_LOCALE}>
            <GraphnosIndex {...initialProps} />
          </IntlProvider>
        </BrowserRouter>
      </Provider>
    )
    expect(spy).not.toHaveBeenCalled()
  })
})
describe('<Graphnos />', () => {
  let store
  beforeAll(() => {
    store = configureStore({}, browserHistory)
  })
  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <BrowserRouter>
          <IntlProvider locale={DEFAULT_LOCALE}>
            <Graphnos {...initialProps} />
          </IntlProvider>
        </BrowserRouter>
      </Provider>,
    )
    expect(firstChild).toMatchSnapshot()
  })

})
