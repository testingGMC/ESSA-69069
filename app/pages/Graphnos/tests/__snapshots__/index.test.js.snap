// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Graphnos /> should render and match the snapshot 1`] = `
<div
  class="graphnos"
>
  <div
    class="four-of-four error-pages"
  >
    <div
      class="content"
    >
      <div
        class="page-not-found-title"
      >
        <span>
          Page not found
        </span>
      </div>
      <div
        class="page-not-found-description"
      >
        <span>
          Sorry, nothing to see here
        </span>
        <br />
        <span>
          Visit our
        </span>
        <a
          href="/dashboard"
        >
           
          <span>
            home page
          </span>
        </a>
        .
      </div>
    </div>
  </div>
</div>
`;
